.Net: [MEVD] Added detailed logging to vector stores

dmytrostruk opened the pull request:
### Motivation and Context

<!-- Thank you for your contribution to the semantic-kernel repo!
Please help reviewers and future users, providing the following information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->

This PR adds detailed logging for vector store operations with information including a collection name, records to upsert, record keys to get and more. 

Example of flow with record upsert and vector search with `LogLevel.Debug`:
```
CollectionExistsAsync invoked. Collection name: skglossary.
CollectionExistsAsync completed. Collection name: skglossary. Collection exists: False
CreateCollectionAsync invoked. Collection name: skglossary.
CreateCollectionAsync completed. Collection name: skglossary.
UpsertAsync invoked. Collection name: skglossary.
UpsertAsync completed. Collection name: skglossary.
UpsertAsync invoked. Collection name: skglossary.
UpsertAsync completed. Collection name: skglossary.
UpsertAsync invoked. Collection name: skglossary.
UpsertAsync completed. Collection name: skglossary.
VectorizedSearchAsync invoked. Collection Name: skglossary.
VectorizedSearchAsync completed. Collection Name: skglossary.
```

Example of flow with Record upsert and vector search with `LogLevel.Trace`:
```
CollectionExistsAsync invoked. Collection name: skglossary.
CollectionExistsAsync completed. Collection name: skglossary. Collection exists: False
CreateCollectionAsync invoked. Collection name: skglossary.
CreateCollectionAsync completed. Collection name: skglossary.
UpsertAsync invoked. Collection name: skglossary.
UpsertAsync completed. Collection name: skglossary. Key: 1.
UpsertAsync invoked. Collection name: skglossary.
UpsertAsync completed. Collection name: skglossary. Key: 2.
UpsertAsync invoked. Collection name: skglossary.
UpsertAsync completed. Collection name: skglossary. Key: 3.
VectorizedSearchAsync invoked. Collection Name: skglossary.
VectorizedSearchAsync completed. Collection Name: skglossary.
```

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

roji commented on dotnet/src/Connectors/VectorData/VectorStorage/LoggingVectorStoreRecordCollection.cs:
```diff
@@ -71,7 +85,12 @@ public Task CreateCollectionIfNotExistsAsync(CancellationToken cancellationToken
-        return LoggingExtensions.RunWithLoggingAsync(
+        if (this._logger.IsEnabled(LogLevel.Trace))
+        {
+            this.LogDelete(this.CollectionName, VectorStoreLoggingExtensions.AsJson(key, this.JsonSerializerOptions));
```
I think this approach of dumping arbitrary values into the log as JSON is problematic in various ways...

* First, any user data (key, metadata) can potentially be sensitive (e.g. personally-identifiable information), and shouldn't be logged by default. We could consider an explicit opt-in to do that if the user wishes - LogLevel.Trace shouldn't be that opt-in though (but rather something that's very specifically about logging sensitive information).
* Using JSON specifically is problematic; for one thing, serializing arbitrary .NET types like this is incompatible with NativeAOT - this is why the suppressions on the AsJson() method are needed. It's also quite heavy to do this (perf-wise) just for dumping a key to the log - given all our currently-supported keys, a simple ToString() would do, I think, if we really want to do this.

To summarize, I'd remove this for now - we can always improve and add something back in the future.

roji replied:
OK, I guess I have a more stringent view here then... I can see that the docs correspond to what you say, but I do still think that a user may enable tracing to get some low-level/perf information, and if that ends up leaking PII that's quite a surprise and a possible security risk. FWIW that's what we do in EF (i.e. a special separate config knob called EnableSensitiveParamterLogging).

I also disilke that you enable trace and then you have **two** messages, duplicating each other in this way, rather than a flag that changes the one message you're getting to contain PII (or not).

But if I'm the only one with this view and given the above docs, I'm OK with that particular aspect of it.

(dumping to JSON like this is another matter which I still don't think we should do)

dmytrostruk replied:
> To summarize, I'd remove this for now - we can always improve and add something back in the future.

Removed JSON serialization for now. 

> I also disilke that you enable trace and then you have two messages, duplicating each other in this way, rather than a flag that changes the one message you're getting to contain PII (or not).

Removed messages duplication.

roji commented on dotnet/src/Connectors/VectorData/VectorStorage/LoggingVectorStoreRecordCollection.cs:
```diff
@@ -42,18 +46,28 @@ public LoggingVectorStoreRecordCollection(IVectorStoreRecordCollection<TKey, TRe
+    public async Task<bool> CollectionExistsAsync(CancellationToken cancellationToken = default)
     {
-        return LoggingExtensions.RunWithLoggingAsync(
+        var collectionExists = await VectorStoreLoggingExtensions.RunWithLoggingAsync(
```
nit: RunWithLoggingAsync can be an extension method over logger if we want.

roji commented on dotnet/src/Connectors/VectorData/VectorStorage/LoggingVectorStoreRecordCollection.cs:
```diff
@@ -89,56 +113,152 @@ public Task DeleteBatchAsync(IEnumerable<TKey> keys, CancellationToken cancellat
+    [LoggerMessage(LogLevel.Trace, "Getting records from '{CollectionName}' with keys: {Keys}")]
+    private partial void LogGetBatch(string collectionName, string keys);
+
+    [LoggerMessage(LogLevel.Trace, "Retrieved record from '{CollectionName}': {Record}")]
```
I'd measure the time (in ms) and log it with each operation, it's probably the first thing we'll get asked for.

dmytrostruk replied:
This is already available in OpenTelemetry decorators in this PR: https://github.com/microsoft/semantic-kernel/pull/11207. 

By the way, the operation duration there is measured in seconds, not milliseconds, based on the convention from OpenTelemetry:
https://opentelemetry.io/docs/specs/semconv/database/database-metrics/#metric-dbclientoperationduration

Based on that, do we still need time measurement in `Logging` decorators as well?

roji replied:
> Based on that, do we still need time measurement in Logging decorators as well?

I think so, yeah - logging and tracing indeed have some functional overlap, but are still two different features: some people only enable logging, and it seems reasonable for those people to expect us to log running times.

Yeah, OTel use seconds for duration as their standard, which I personally find a big odd; IIRC that can still be a fraction (e.g. 1.255s). I'm not sure what the appropriate unit of measure here would be (how much does embedding generation typically take?), but I'd err towards milliseconds rather than seconds...

roji commented on dotnet/src/Connectors/VectorData/VectorStorage/LoggingVectorStoreRecordCollection.cs:
```diff
@@ -89,56 +113,152 @@ public Task DeleteBatchAsync(IEnumerable<TKey> keys, CancellationToken cancellat
+
+    #region private
+
+    [LoggerMessage(LogLevel.Trace, "Collection '{CollectionName}' exists: {CollectionExists}")]
```
Consider adding EventIds here for each event, to allow consumers to e.g. filter on them.

dmytrostruk replied:
Added Event ID for each log message and put all Event IDs in a separate internal constant class, see `LoggingEventIds`. Although I'm not sure if ID pattern is good, please let me know what you think about it:
![image](https://github.com/user-attachments/assets/90d922e6-4d69-46ea-b799-ca7aa353880d)

roji replied:
I think that looks pretty good! As a total minor nit, maybe `Starting` is a bit more accurate/clear than `Invoked`, since nothing has actually been invoked yet when that log message is emitted.

roji commented on dotnet/src/Connectors/VectorData/Extensions/VectorStoreLoggingExtensions.cs:
```diff
@@ -123,6 +127,81 @@ internal static async IAsyncEnumerable<TResult> RunWithLoggingAsync<TResult>(
         }
     }
 
+    [UnconditionalSuppressMessage("AotAnalysis", "IL3050", Justification = "DefaultJsonTypeInfoResolver is only used when reflection-based serialization is enabled")]
```
I'm not sure I understand the suppression here: the JsonSerializer.Serialize() as used here is indeed incompatible with NativeAOT, no?

dmytrostruk replied:
The serialization happens only when `JsonSerializer.IsReflectionEnabledByDefault` is true. Based on my understanding, this logic is compatible with NativeAOT, but suppression message is still required, since the compiler doesn't understand the usage of `JsonSerializer.IsReflectionEnabledByDefault`. This logic was taken from MEAI:
https://github.com/dotnet/extensions/blob/56df0c4f4933909367536413dcf9c16520a37f82/src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Defaults.cs#L43

But in all cases, I completely removed JSON serialization in this version of logging, and as you mentioned above, we can always extend it later if needed.

roji commented on dotnet/src/Connectors/VectorData/Extensions/VectorStoreLoggingExtensions.cs:
```diff
@@ -134,4 +213,6 @@ internal static async IAsyncEnumerable<TResult> RunWithLoggingAsync<TResult>(
     [LoggerMessage(LogLevel.Error, "{OperationName} failed.")]
     private static partial void LogInvocationFailed(this ILogger logger, string operationName, Exception exception);
+
+    #endregion
```
As the default logging experience, I'd expect to only see `Collection 'Foo' created in 3ms`. In other words, I think the "invoked" message (which should probably be named "invoking" as it's logged before the operation) should be in a lower level than the Completed message (so you only get the Completed message).

dmytrostruk replied:
I tried to be consistent with MEAI logging naming, but I'm happy to adjust it based on our preferences:
https://github.com/dotnet/extensions/blob/56df0c4f4933909367536413dcf9c16520a37f82/src/Libraries/Microsoft.Extensions.AI/ChatCompletion/LoggingChatClient.cs#L173-L195

roji replied:
OK, thanks - I agree it makes sense to be consistent with MEAI here. @SteveSandersonMS what do you think, should the default experience include both before (invoked) and after (completed) messages, or just completed (and have invoked at trace)?

roji reviewed (changes requested):
Hey @dmytrostruk, here's a preliminary review; I didn't go too far so that we can first align on the general principles here etc.

In general, there seem to be two different logging approaches implemented here:

* One on LogLevel.Trace only, using the source generator (`[LoggerMessage]`). Here the messages are customized (e.g. CreateCollection logs the name of the collection created).
* One via a general VectorStoreLoggingExtensions utility, emitting a log message before, and another after (success/failure/cancellation). Here the messages are only "OperationName invoked/completed", no customization.

One problem here is that the default, non-trace logging gives no information - one expects to see which collection was created with the event (`CreateCollection completed` is, well, a bit incomplete). It seems that one has to activate LogLevel.Trace to get that, and at that point they get duplicate messages. In other words, I'd expect just a single system here that's informative (contains any information relevant to the operation).

On a related note, it's best practice to include Event IDs when logging, so that users can filter for specific events etc. Doing this - and doing it efficiently - probably means that you'd need to have the try/catch/finally that's currently in VectorStoreLoggingExtensions simply in the methods themselves, and call distinct, source-generated methods each time (CreateCollectionString(), CreateCollectionDone(), CreateCollectionFailed(), CreateCollectionCancelled(), each with its own message, event ID, etc.).

Also, see comments below about dumping user values as JSON into the log message - I think this is problematic.

Finally, I can't shake the feeling that the explosion of types (all the wrappers in VectorSearch and VectorStorage) indicates we're going in a problematic direction, for this particular abstraction. I can't really say anything more concrete than that, and I'm OK with merging something like this, but it feels a bit off.

dmytrostruk commented:
@roji Thanks for PR review!

> One problem here is that the default, non-trace logging gives no information - one expects to see which collection was created with the event (CreateCollection completed is, well, a bit incomplete). It seems that one has to activate LogLevel.Trace to get that, and at that point they get duplicate messages. In other words, I'd expect just a single system here that's informative (contains any information relevant to the operation).

I removed message duplication and now there is only one message based on `LogLevel` value.

> On a related note, it's best practice to include Event IDs when logging, so that users can filter for specific events etc. Doing this - and doing it efficiently - probably means that you'd need to have the try/catch/finally that's currently in VectorStoreLoggingExtensions simply in the methods themselves, and call distinct, source-generated methods each time (CreateCollectionString(), CreateCollectionDone(), CreateCollectionFailed(), CreateCollectionCancelled(), each with its own message, event ID, etc.).

I updated the code to have a dedicated log message for each operation and its possible outcome (i.e. completed/canceled/failed) and added a unique Event ID for each message.

> Also, see comments below about dumping user values as JSON into the log message - I think this is problematic.

I removed JSON serialization for now.

roji reviewed (commented):
> Finally, I can't shake the feeling that the explosion of types (all the wrappers in VectorSearch and VectorStorage) indicates we're going in a problematic direction, for this particular abstraction. I can't really say anything more concrete than that, and I'm OK with merging something like this, but it feels a bit off.

So on this... I had a chat with @stephentoub and thought about this some more.. Here are some ideas.

In general, I think we should think of the search interfaces (e.g. IVectorizedSearch, IKeywordHybridSearch) as **views** over a collection; somewhere, at some point in the user's application, there's going to be an actual concrete IVectorStoreRecordCollection, and the views just point to that. If we look at it that way, then it makes sense to apply all decorators on the collection, and **then** take interfaces which point to that. This would mean the following:

* We shouldn't need any wrappers for IVectorizedSearch, since IVectorStoreRecordCollection extends it.
    * It could be said that in theory, someone might have e.g. an implementation of IVectorizedSearch that **doesn't** point to an IVectorStoreRecordCollection, and that because of that a builder+wrapper specifically for IVectorizedSearch may be needed. I think that probably quite contrived, and would do this only if we have confirmed user cases which make this necessary; we can always add this in the future if necessary (no breaking change).
* IVectorizedTextSearch will likely go the same way, since we're planning to have all collections support embedding generation as part of [#10492](https://github.com/microsoft/semantic-kernel/issues/10492). So IVectorizedSearch (or whatever we call it, maybe just IVectorSearch) would likely just have both SearchEmbeddingAsync (corresponding to today's VectorizedSearch) and SearchAsync (corresponding to today's VectorizableSearch).
* That leaves hybrid search, which really is an optional interface (IVectorStoreRecordCollection doesn't extend it).
    * We already have a problem of how a user actually gets LoggingKeywordHybridSearch registered in DI. Doing e.g. `.AddWeaviateVectorStoreRecordCollection()` registered IKeywordHybridSearch in DI (when the connector supports it), but does `.AddWeaviateVectorStoreRecordCollection().UseLogging()` register a LoggingKeywordHybridSearch for the user? I might have missed it, but I don't think so (only the collection decorator seems to get registered).
    * We could have UseLogging() internally check if the IVectorStoreRecordCollection implements IKeywordHybridSearch, and if so, wrap that and register the wrapper in DI. This way, the user automatically gets everything registered without having to do extra manual steps.
    * If that works, then we don't need a builder pipeline for IKeywordHybridSearch, nor do we need to expose the decorator publicly (in general, I'm not sure we need to expose any of them publicly).

If the above makes sense, if means we can pretty much get rid of [the entire builders/wrappers over the search interfaces](https://github.com/microsoft/semantic-kernel/tree/9926e5cfddb522e66b5fa5df996f6d62f9b87467/dotnet/src/Connectors/VectorData/VectorSearch), and this becomes much more palatable, requiring the user to do much less and exposing less public APIs.

What do you think?

dmytrostruk commented:
This will be added later.
